Networking test rewrite

bprashanth opened the pull request:
Decomposes the old kubeproxy tests into (tcp, udp) tests for each of the following: 
- intra-pod
- node-pod
- pod-Service
- node-Service
- endpoint-Service
- delete endpoints, confirm unreachability
- delete nodeport, confirm unreachability
- kube-proxy /proxymode, /healthz

Also gets rid of the old network conformance test that used apiserver proxy to check reported peer count of a webserver in a container (the netexec pod used in this test does the same thing without apiserver proxy). 

Fixes https://github.com/kubernetes/kubernetes/issues/26490, https://github.com/kubernetes/kubernetes/issues/14204

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31559)

<!-- Reviewable:end -->

bprashanth committed:
Networking test rewrite

bprashanth commented on test/e2e/networking.go:
```diff
@@ -79,235 +75,172 @@ var _ = framework.KubeDescribe("Networking", func() {
-			err = f.WaitForPodRunning(podName)
-			Expect(err).NotTo(HaveOccurred())
-		}
+	// TODO: Remove [Slow] when this has had enough bake time to prove presubmit worthiness.
```
This decomposes a SLOW test (1 Describe + 1 It ~20m) into 2 Describes + 15 It blocks each taking < 5m (but still ~20m serially). @kubernetes/sig-testing running this in presubmit should be fine because It runs in parallel right?

girishkalele commented:
@bprashanth Should I hold off reviewing this till you remove the wip in the title ?

bprashanth commented:
@girishkalele open for review, just running tests in a loop to weed out some flakes but this mostly won't change drastically

bprashanth commented:
I'm running into flakes in my overnight runs because of https://github.com/kubernetes/kubernetes/issues/31630, I think, otherwise this should be ok

girishkalele commented:
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/e2e/networking_utils.go, line 46 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQN0-BXEjh5M0XOCMbr:-KQN0-BXEjh5M0XOCMbs:be68teh) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L46)):*
> ```Go
> 	clusterUdpPort          = 90
> 	loadBalancerHttpPort    = 100
> 	netexecImageName        = "gcr.io/google_containers/netexec:1.5"
> ```

FYI - I have pushed 1.6 recently, with a new clientip endpoint.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/e2e/networking_utils.go, line 340 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQN0pYjMadEhO89Ge69:-KQN0pYjMadEhO89Ge6A:b-iel38e) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L340)):*
> ```Go
> 			Type: api.ServiceTypeNodePort,
> 			Ports: []api.ServicePort{
> 				{Port: clusterHttpPort, Name: "http", Protocol: api.ProtocolTCP, TargetPort: intstr.FromInt(endpointHttpPort)},
> ```

Do we want to be able to specify session affinity as an input to this method ?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Reviewed 1 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/e2e/networking.go, line 194 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQNF_F-pmVepvmyI1YM:-KQNF_F-pmVepvmyI1YN:b-5nyrn4) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking.go#L194)):*
> ```Go
> 			config.dialFromTestContainer("udp", config.clusterIP, clusterUdpPort, config.maxTries, 0, config.endpointHostnames())
> 
> 			config.deleteNetProxyPod()
> ```

Are we deleting this twice by mistake ?
Are these It() clauses going to run in parallel/concurrently ?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/e2e/networking.go, line 194 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQNF_F-pmVepvmyI1YM:-KQNKImdA233-PtFVUb6:b-himzyk) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking.go#L194)):*
<details><summary><i>Previously, girishkalele (Girish Kalele) wrote…</i></summary>
> Are we deleting this twice by mistake ?
> Are these It() clauses going to run in parallel/concurrently ?
</details>
Never mind, I see a new object created in every It() clause.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/e2e/networking_utils.go, line 51 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQNLO-o6NR80l77Xgg1:-KQNLO-o6NR80l77Xgg2:b-3rf36j) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L51)):*
> ```Go
> 	hostTestPodName         = "host-test-container-pod"
> 	nodePortServiceName     = "node-port-service"
> 	loadBalancerServiceName = "load-balancer-service"
> ```

Is this loadBalancerServiceName used ? And the loadBalancerHttpPort ? I saw only services of type NodePort.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
Reviewed 1 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
I have minor nits about the loadbalancer vars, otherwise LGTM

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-:-KQNMPVUIpO0KjlwereO:b-t0b7vd)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-:-KQNMZCL6AcYp4ilEXYu:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[test/e2e/networking_utils.go, line 46 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQN0-BXEjh5M0XOCMbr:-KQNMbNv92nOpVe79uq6:b-e5whnr) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L46)):*
<details><summary><i>Previously, girishkalele (Girish Kalele) wrote…</i></summary>
> FYI - I have pushed 1.6 recently, with a new clientip endpoint.
</details>
will change

---

*[test/e2e/networking_utils.go, line 51 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQNLO-o6NR80l77Xgg1:-KQNLvrOzL8RQdpHahaZ:b-an8bz4) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L51)):*
<details><summary><i>Previously, girishkalele (Girish Kalele) wrote…</i></summary>
> Is this loadBalancerServiceName used ? And the loadBalancerHttpPort ? I saw only services of type NodePort.
</details>
no that's from the old file, will remove dead code

---

*[test/e2e/networking_utils.go, line 340 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQN0pYjMadEhO89Ge69:-KQNM48lIPxMZgilEMHS:bm638po) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L340)):*
<details><summary><i>Previously, girishkalele (Girish Kalele) wrote…</i></summary>
> Do we want to be able to specify session affinity as an input to this method ?
</details>
yeah we should write another test that sets affinity, then curls N times from pod/node etc and confirms that we don't see any other endpoint. Can I file a bug to follow up? 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
Going to upload another pr to address comments and test overnight with @freehan s fix, applying do-not-merge for now

bprashanth commented:
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/e2e/networking_utils.go, line 51 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQNLO-o6NR80l77Xgg1:-KQRJx_8o7f5Qw4Kfhnu:b-hfpv0l) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L51)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> no that's from the old file, will remove dead code
</details>
removed

---

*[test/e2e/networking_utils.go, line 340 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQN0pYjMadEhO89Ge69:-KQRJyxOtFNAXSO5aDpy:byopq0i) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L340)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> yeah we should write another test that sets affinity, then curls N times from pod/node etc and confirms that we don't see any other endpoint. Can I file a bug to follow up? 
</details>
Added todo and link to https://github.com/kubernetes/kubernetes/issues/31712

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
Ran overnight with CNI plugin built with https://github.com/containernetworking/cni/pull/292 and got green runs. Waiting for a new cni build to push this out, because otherwise some `[Conformance]` tests will start flaking in presubmit.

wojtek-t commented:
@bprashanth - I didn't take any closer look into this PR, but can you please ensure that tests that you're rewriting will also work in large clusters (2000-nodes) and will not take ages there?

bprashanth commented:
Since https://github.com/kubernetes/kubernetes/pull/31801 just went in i plan to give this another overnight run and lgtm tomorrow if things look ok.
@wojtek-t these tests should be scale invariant, unless there's eg: something putting so much load on every kubeproxy in a large cluster that it takes ages to notice a readiness probe. 

I think the long times would be observed on loadbalancer tests because it takes longer to update  resources in the cloud/health check every node etc, @girishkalele fyi. I don't know how to address that though, except for opening an internal bug.

wojtek-t commented:
>
I think the long times would be observed on loadbalancer tests because it takes longer to update resources in the cloud/health check every node etc, @girishkalele fyi. I don't know how to address that though, except for opening an internal bug.

Yes - that one I'm aware of and this is the only test that I accept that we can't do anything reasonable about it and it can run for some time (the currently existing test takes ~1 hour - if the new one won't be significantly longer in 2000-node cluster, then it's fine).

marun commented:
Is there a reason the granular pod-pod connectivity checks (including across nodes) were removed?  These were added as a sanity check for 3rd party network plugins and I think they are still useful.

bprashanth commented:
That should already be happening by default in all the "Granular Checks: Pods" tests (pod-pod http/udp, node-pod http/udp). We create 1 pod per node in the cluster, then try to access that pod using http/udp from another container (pod-pod) and another node (ssh). So as long as we're looping over all endpoints we're testing cross node and same node.

bprashanth commented:
@freehan @girishkalele ptal, rebase and finished local testing so I think this should be good to go now that the CNI fix is in.

girishkalele commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/e2e/networking_utils.go, line 46 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQN0-BXEjh5M0XOCMbr:-KR5OT2vY5luoJ4oeMZC:b-x9mcr3) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L46)):*
<details><summary><i>Previously, bprashanth (Prashanth B) wrote…</i></summary>
> will change
</details>
Think you missed the change to 1.6, not a big deal, the only addition was the /clientip endpoint.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

girishkalele commented:
@freehan LGTM

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-:-KR5OozHLo3cXykWBwBv:bww4j3a)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[test/e2e/networking_utils.go, line 46 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559#-KQN0-BXEjh5M0XOCMbr:-KR5RJlO6lnZaNxD0As4:budx1hw) ([raw file](https://github.com/kubernetes/kubernetes/blob/f6e7d19226c3871072491352bc931374727c1950/test/e2e/networking_utils.go#L46)):*
<details><summary><i>Previously, girishkalele (Girish Kalele) wrote…</i></summary>
> Think you missed the change to 1.6, not a big deal, the only addition was the /clientip endpoint.
</details>
oh, sorry, will bump it up with the next clarity pr i send (there will surely be one on the first flake) since this passed e2es already and I've tested it locally

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31559)*
<!-- Sent from Reviewable.io -->

bprashanth commented:
Bumping this to a p2 because it deals with flaky tests

wojtek-t commented:
@bprashanth - those tests completely broke large clusters; basically, all of them are super big (there are ~20 tests, each of them creating 2000 pods (in 2000-node cluster) and doing a bunch of operations). Running them in parallel only makes things worse.

That said, the test for load balancer, is not the most expensive one now.

I will send a PR to limit the number of pods within a service to some constant number.

wojtek-t commented:
https://github.com/kubernetes/kubernetes/pull/32659 is the PR that is addressing it

bprashanth commented:
@wojtek-t forgot it was creating an endpoint per node, apologies, thanks for the fix
